Objective-C/CBLRemoteDatabase.mm

```diff
@@ -0,0 +1,219 @@
+            .callbackContext    = (__bridge void*)self,
+        };
+        C4Error c4err = {};
+        _client = c4client_new(&params, &c4err);
+    }
+    
+    return self;
+}
+
+- (void) dealloc {
+    c4client_free(_client);
+    
+    _client = nil;
+}
+
+#pragma mark - Stop
+
+- (void) stop {
+    c4client_stop(_client);
+}
+
+#pragma mark - Get Document
+
+static void getDocumentCallback(C4ConnectedClient* c4client,
+                                  const C4DocResponse* doc,
+                                  C4Error* err,
+                                  void* context) {
+    ConnectedClientGetDocumentContext* ctx = (__bridge ConnectedClientGetDocumentContext*)context;
+    CBLDocument* cblDoc = nil;
+    NSError* error = nil;
+    if (err == nil || err->code == 0) {
+        if (doc->deleted) {
+            cblDoc = nil;
+            error = [NSError errorWithDomain: CBLErrorDomain
+                                        code: CBLErrorNotFound
+                                    userInfo: nil];
+                
+        } else {
+            FLSliceResult res = FLSliceResult(alloc_slice(doc->body));
+            cblDoc = [[CBLDocument alloc] initWithDocumentID: slice2string(doc->docID)
+                                                  revisionID: slice2string(doc->revID)
+                                                        body: res];
+            FLSliceResult_Release(res);
+        }
+    } else if (err) {
+        convertError(*err, &error);
+    }
+    
+    dispatch_async(ctx.remoteDB.dispatchQueue, ^{
+        ctx.docGetCompletion(cblDoc, error);
+        [ctx.remoteDB.contexts removeObject: ctx];
+    });
+}
+
+- (void) documentWithID: (NSString*)identifier
+             completion: (void (^)(CBLDocument*, NSError*))completion {
+    CBLStringBytes docID(identifier);
+    _c4err = {};
+    ConnectedClientGetDocumentContext* ctx = [[ConnectedClientGetDocumentContext alloc] initWithRemoteDB: self
+                                                                                            completion: completion];
+    [_contexts addObject: ctx];
+    c4client_getDoc(_client, docID, nullslice, nullslice, true, &getDocumentCallback, (__bridge void*)ctx, &_c4err);
+}
+
+#pragma mark - Save
+
+- (void) saveDocument: (CBLMutableDocument *)document
+           completion: (void (^)(CBLDocument*, NSError*))completion {
+    [self saveDocument: document asDeletion: NO completion: completion];
+}
+
+- (void) deleteDocument: (CBLDocument *)document
+             completion: (void (^)(CBLDocument*, NSError*))completion {
+    [self saveDocument: document asDeletion: YES completion: completion];
+}
+
+static void updateDocumentCallback(C4ConnectedClient* c4client, C4HeapSlice newRevID, C4Error* err, void *context) {
+    ConnectedClientUpdateDocumentContext* ctx = (__bridge ConnectedClientUpdateDocumentContext*)context;
+    BOOL success = err == nil || err->code == 0;
+    NSError* error = nil;
+    if (!success && err) {
```

Comment:
Maybe just if (err && err-> code != 0)?

Reply:
Also is there a case of err != nil && err-> code == 0?

Reply:
I guess, `if (err && err-> code != 0)` this should be fine. I might have over thought about this